pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/typeinference/TypeInferenceResolver.java

```diff
@@ -31,6 +32,38 @@ private TypeInferenceResolver() {
 
     }
 
+    public static List<JavaTypeDefinition> inferTypes(List<Constraint> constraints, List<Bound> bounds,
+                                                      List<Variable> variables) {
+
+        List<Bound> newBounds = new ArrayList<>();
+        while (!constraints.isEmpty()) {
+            List<BoundOrConstraint> reduceResult = constraints.get(constraints.size() - 1).reduce();
+            constraints.remove(constraints.size() - 1);
```

Comment:
these 2 lines could be condensed. `remove` returns the removed element, so you could directly write: `List<BoundOrConstraint> reduceResult = constraints.remove(constraints.size() - 1).reduce();`